cpp/onedal/array.hpp

```diff
@@ -32,54 +34,155 @@ class array {
   friend array<Y> const_array_cast(const array<U>&);
 
   public:
-    explicit array(std::int64_t element_count)
-        : data_(new T[element_count]), // TODO: Use custom oneDAL allocator
-          size_(element_count) {}
-
-    explicit array(const detail::shared<T>& data, std::int64_t size)
-        : data_(data),
-          size_(size) {}
+    array()
+        : data_owned_ptr_(nullptr),
+          size_(0),
+          capacity_(0) {}
+
+    explicit array(std::int64_t element_count, T element = {})
+        : data_owned_ptr_(new T[element_count], // TODO: Use custom oneDAL allocator
+                [](T* obj){ delete[] obj; }),
+          data_(data_owned_ptr_.get()),
+          size_(element_count),
+          capacity_(element_count) {
+        for (std::int64_t i = 0; i < size_; i++) {
+            (*this)[i] = element;
+        }
+    }
 
     template <typename Deleter>
     explicit array(T* data, std::int64_t size, Deleter&& deleter)
-        : data_(data, deleter),
-          size_(size) {}
+        : data_owned_ptr_(data, std::forward<Deleter>(deleter)),
+          data_(data),
+          size_(size),
+          capacity_(size) {}
 
     template <typename U>
     array(const array<U>& other)
-        : data_(other.data_),
-          size_(other.size_) {}
+        : data_owned_ptr_(other.data_owned_ptr_),
+          data_(other.data_),
+          size_(other.size_),
+          capacity_(other.capacity_) {}
+
+    T* get_mutable_data() const {
+        return std::get<T*>(data_);
```

Comment:
Need to think about exception here. As far as I understand, we'll need to convert `bad_variant_access` to some dal-specific exception.

Reply:
Yes, exactly. I am not converting this right now because we have no exceptions on our side at now 😃
Will place TODO here not to forget this place.